api/envoy/config/listener/v3/listener.proto

```diff
@@ -35,7 +37,7 @@ message ListenerCollection {
   repeated xds.core.v3.CollectionEntry entries = 1;
 }
 
-// [#next-free-field: 31]
+// [#next-free-field: 33]
```

Comment:
This doesn't need to block the PR, but if we wanted to really prove out how usable the API is, I think we could do it without implementing it in Envoy. Snow has a golang implementation of the generic matcher. Istio already has an extensive test suite for filter chain matching, based on a golang implementation of the current FCM logic. So if we were able to implement the new matcher in Istio (which presumably we need to do anyways after this merges), we could get full test coverage as well before committing to the Envoy implementation.

All of that may be harder than just doing the implementation though, just throwing out the idea.

Reply:
Generic matcher does not have custom match implemented in either envoy or golang, so it's not really realistic at this stage.

Reply:
I was referring to https://github.com/envoyproxy/go-control-plane/pull/511. I guess would need to add all the custom matchers though